runtime/lua/vim/diagnostic.lua

```diff
@@ -1857,16 +1857,37 @@ function M.open_float(opts, ...)
   if scope == 'line' then
     --- @param d vim.Diagnostic
     diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum and lnum <= d.end_lnum
+      if lnum < d.lnum or lnum > d.end_lnum then
+        return false
+      end
+      if d.lnum == d.end_lnum then
+        -- if diagnostic only spans one line, always show
+        -- allows showing diagnostics with zero width (where lnum == end_lnum and col == end_col)
+        return true
+      else
+        -- end-exclusive range, don't show if on end_lnum and end_col == 0
+        return lnum ~= d.end_lnum or d.end_col ~= 0
+      end
     end, diagnostics)
   elseif scope == 'cursor' then
-    -- LSP servers can send diagnostics with `end_col` past the length of the line
-    local line_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
+    -- LSP servers can send diagnostics with `col` past the end of the line
+    local lnum_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
```

Comment:
This comment changes `end_col` to `col`. Was this intentional? Was the old comment incorrect?

If the server can send both `col` and `end_col` past the end of the line we should mention both here.

Do you have an example of a server sending `col` (not `end_col`) past the end of the line? That seems very peculiar to me.

Reply:
This was intentional yes - the issue that was meant to fix (#16062) gives this example, where `col` is past the end of the line:

> The easiest way to reproduce the issue is to directly create diagnostics using `vim.diagnostic.set`. Open a buffer with the following content:
> 
> ```
> hello
> hello
> ```
> 
> Then run the following Lua code to set a diagnostic at the end of the first line:
> 
> ```lua
> vim.diagnostic.set( vim.api.nvim_create_namespace("test"), 0, { { lnum = 0, end_lnum = 1, col = 5, end_col = 0, message = "test" } })
> ```
> 
> (Note that the range is meant to specifically denote the end of the first line, not including the final `o` in `hello`. More on this below.)

Everything is zero-based, so `col = 5` is past the end of the line. `end_col` can also be past the end of the line, but that doesn't need special treatment - the condition `col < d.end_col` will always be true and the float will just _always_ be shown. `col` past the end of the line does need special treatment, because in that case the condition `col < d.col` will always be true and the float will _never_ be shown.

As for an actual example, the #16062 mentions `tsserver`, but "didn't have an example handy", and neither do I :( I can try and find one if you'd like.